teps/0138-improve-tekton-pipeline-feature-flags.md

```diff
@@ -0,0 +1,73 @@
+---
+status: proposed
+title: Improve Tekton Pipeline Feature Flags
+creation-date: '2023-07-07'
+last-updated: '2023-07-07'
+authors:
+- '@JeromeJu'
+- '@chitrangpatel'
+- '@lbernick'
+---
+
+# TEP-0138: Improve Tekton Pipeline Feature Flags
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+
+## Summary
+
+This document proposes updating Tekton Pipelines' feature flags design, as originally proposed in [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md), to address problems related to the coupling API versioning and feature versioning, and to provide a better experience for cluster operators.
+
+## Motivation
+
+Currently there are two design decisions from [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md), grouped feature flags and coupled feature version and apiVersion, that led to challenges for cluster operators, users and maintainers for the usage of existent feature flags.
+
+### Grouped feature flags
+
+New features in Tekton start at "`alpha`" stability and can be promoted to `beta` or "`stable`". Today, we have a single feature flag, `enable-api-fields`, that enables all fields in the Tekton API supporting features at a certain stability level. `enable-api-fields` can be set to "`alpha`", `beta`, or "`stable`", specifying the minimum stability level for any fields used in Tekton APIs on that cluster. In addition, features that aren't controlled by fields in Tekton APIs have their own feature flags. For more information, see [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md). This approach has led to the following pain points: 
+
+- Cluster operators can't enable individual `alpha` or `beta` features controlled by API fields.
+  - Since features at a lower stability level tend to have more bugs, cluster operators and vendors may want to limit usage of lower-stability features they do not need, while still being able to enable individual features for specific use cases.
```

Comment:
Enabling all alpha features does not mean that all those features are used as well, and vendors could still document that only certain features are enabled, but I understand that that is not the same as preventing a feature from being used.

One note is that the controller code usually is not feature flag gated either, but certain code paths should only be enabled if the corresponding API fields are used, and we control that via validation.
Activating all alpha flags does activate some extra code paths in the validation code which might have some unwanted side effects, so probably that this is the problem we would like to avoid by using individual flags.

Reply:
Getting back to this comment, as a TODO once this TEP proposal is merged is to have clarifications in the API-compatibility policy that all per-features flags are supposed to be independent from each other. I hope this could resolve the concerns here:)